File: cpp/crsce/compress/arguments.cpp
/*
 * Crsce/compress/arguments.cpp
 * (c) 2023 Sam Caldwell.  See LICENSE.txt
 */
#include "arguments.h"
#include "../../../projects/cpp/misc/trim.cpp"


/*
 * Parse the command line arguments and configure
 * the Argument class.
 */
void Argument::Parse(int argc, char *argv[]) {
    std::string currArg;
    std::string prevArg;
    bool expectValue = false;

    for (int i = 0; i < argc; i++) {

        currArg = misc::trim(argv[i]);

        if (expectValue) {
            if (currArg.substr(0, 1) == argShort) {
                throw std::runtime_error("expected value");
            }
            if (prevArg == argInput) {
                this->__in = currArg;
                prevArg = emptyString;
            }
            if (prevArg == argOutput) {
                this->__out = currArg;
                prevArg = emptyString;
            }

        } else {

            if ((currArg == argInput) || (currArg == argOutput)) {
                expectValue = true;
                prevArg = currArg;
            } else {
                expectValue = false;
                if (currArg == argColor) {
                    this->__color = true;
                    continue;
                }
                if (currArg == argDebug) {
                    this->__debug = true;
                    continue;
                }
                if (currArg == argHelp) {
                    this->__help = true;
                    continue;
                }
            }
        }
    }
    if (this->__in == emptyString) {
        throw std::runtime_error("Missing --in <string> value");
    }
    if (this->__in == emptyString) {
        throw std::runtime_error("Missing --out <string> value");
    }
};

File: go/crsce/compress/lib/Arguments.go
package crsce

import (
	"flag"
	"fmt"
	"github.com/sam-caldwell/monorepo/go/exit"
	"github.com/sam-caldwell/monorepo/go/fs/file"
	"github.com/sam-caldwell/monorepo/go/misc/words"
	"strings"
)

const (
	commandUsage = `

crsce-compress --in <input_file> --out <output_file> [--bs <blocksize>]

`
)

const (
	defaultBlockSize = 1048576 //bytes
	s                = 8192    // ceil(sqrt(1048576)) * 8 (bits)
)

// Arguments - Struct representing parsed command-line args.
type Arguments struct {
	In        string
	Out       string
	BlockSize uint
}

// GetArgs - Parse cli arguments and validate the values.
func (args *Arguments) GetArgs() {
	const (
		MinSize = 1024
		MaxSize = 65535
	)
	shortHelp := flag.Bool("-h", false, "Display Usage")
	longHelp := flag.Bool("--help", false, "Display Usage")
	inFile := flag.String("--in", words.EmptyString, "input file")
	outFile := flag.String("--out", words.EmptyString, "output file")
	blockSize := flag.Uint("--bs", defaultBlockSize, "")
	flag.Parse()

	exit.OnCondition(*shortHelp || *longHelp, exit.Success, "usage:", commandUsage)

	args.In = *inFile
	exit.OnCondition(strings.TrimSpace(args.In) == "", exit.InvalidInput, "input file cannot be empty", commandUsage)
	exit.OnCondition(!file.Exists(args.In), exit.InvalidInput, "input file does not exist.", commandUsage)

	args.Out = *outFile
	exit.OnCondition(strings.TrimSpace(args.Out) == "", exit.InvalidInput, "output file cannot be empty", commandUsage)
	exit.OnCondition(file.Exists(args.Out), exit.InvalidInput, "output file should not exist.", commandUsage)

	args.BlockSize = *blockSize
	exit.OnCondition(args.BlockSize <= MinSize, exit.InvalidInput, fmt.Sprintf("minimum block size %d", MinSize), commandUsage)
	exit.OnCondition(args.BlockSize > MaxSize, exit.InvalidInput, fmt.Sprintf("minimum block size %d", MinSize), commandUsage)
}
